packages/loopring_v3/contracts/aux/fast-withdrawal/FastWithdrawalLiquidityProvider.sol

```diff
@@ -81,21 +114,97 @@ contract FastWithdrawalLiquidityProvider is ReentrancyGuard, OwnerManagable
         agent.executeFastWithdrawals{value: value}(withdrawals);
     }
 
-    // Allows the LP to transfer funds back out of this contract.
-    function drain(
-        address to,
+    function deposit(
         address token,
-        uint    amount
+        uint amount
         )
         external
         nonReentrant
-        onlyOwner
     {
+        require(amount > 0, "ZERO_AMOUNT");
+        uint _balance = 0;
         if (token == address(0)) {
-            to.sendETHAndVerify(amount, gasleft()); // ETH
+            _balance = address(this).balance;
+            address(this).sendETHAndVerify(amount, gasleft()); // ETH
         } else {
-            token.safeTransferAndVerify(to, amount);  // ERC20 token
+            _balance = ERC20(token).balanceOf(address(this));
+            token.safeTransferAndVerify(address(this), amount);  // ERC20 token
         }
+        userTokenBalances[msg.sender][token] = userTokenBalances[msg.sender][token].add(amount);
+
+        uint _sharesToAdd = 0;
+        uint _totalShares = tokenShares[token];
+        if (_totalShares == 0) {
+            _sharesToAdd = 100 * SHARE_BASE;
+        } else {
+            _sharesToAdd = _totalShares.mul(amount) / _balance;
```

Comment:
`_balance` is the current balance in the contract, but there is a very good chance that some funds are currently loaned out for fast withdrawals, so `_balance` is not the actual total balance deposited (+ fees earned). I think that would make this calculation give the depositer more shares than he should have (in the worst case when almost all funds are loaned out this would print almost an infinite number of shares for the depositor).

The fact that the idea is to earn fees directly on this contract makes it very hard to track if funds are currently loaned out or not (without making the fast withdrawals more expensive by storing extra state onchain, which wouldn't make sense I think). The only solution I currently see is that `_sharesToAdd = amount` in all cases, so 1:1 mapping, which at least guarantees that the amount deposited is correctly tracked. But that still doesn't solve how fees can be paid out directly (only the correct share of the fees can be deduced when there are no loans for the token). Perhaps it is good enough in the contract to check for `balance >= totalDeposited` as the check for no open loans, and distribute the difference between depositors, but this would not always be correct of course.

Reply:
good point, I didn't think of it. I think this part of logic should be controlled by the relayer, not the smart contract.